From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Fri, 23 Dec 2022 20:42:50 +0100
Subject: [PATCH] Reduce block destruction packet allocations

License: GPL-3.0-only (https://www.gnu.org/licenses/gpl-3.0.html)
Gale - https://galemc.org

This patch is based on the following patch:
"Cache block break animation packet"
By: VytskaLT <VytskaLT@protonmail.com>
As part of: SportPaper (https://github.com/Electroid/SportPaper)
Licensed under: GPL-3.0-only (https://www.gnu.org/licenses/gpl-3.0.html)

diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 5881e6f68ad6108c35ec4b61c338b0f3572b366b..c0c370edcdd335bb3ea9c9142db3afa4465f5118 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1817,6 +1817,11 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
     @Override
     public void destroyBlockProgress(final int id, final BlockPos blockPos, final int progress) {
+        // Gale start - SportPaper - reduce block destruction packet allocations
+        var players = this.server.getPlayerList().getPlayers();
+        if (players.isEmpty()) return;
+        ClientboundBlockDestructionPacket packet = new ClientboundBlockDestructionPacket(id, blockPos, progress);
+        // Gale end - SportPaper - reduce block destruction packet allocations
         // CraftBukkit start
         Player breakerPlayer = null;
         Entity entity = this.getEntity(id);
@@ -1833,7 +1838,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                 .callEvent();
         }
         // Paper end - Add BlockBreakProgressUpdateEvent
-        for (ServerPlayer player : this.server.getPlayerList().getPlayers()) {
+        for (ServerPlayer player : players) { // Gale - SportPaper - reduce block destruction packet allocations
             if (player.level() == this && player.getId() != id) {
                 double xd = blockPos.getX() - player.getX();
                 double yd = blockPos.getY() - player.getY();
@@ -1844,7 +1849,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                         continue;
                     }
                     // CraftBukkit end
-                    player.connection.send(new ClientboundBlockDestructionPacket(id, blockPos, progress));
+                    player.connection.send(packet); // Gale - SportPaper - reduce block destruction packet allocations
                 }
             }
         }
